File: ios/LiulianUI/Sources/LiulianUI/LiulianCard.swift
import SwiftUI

/// LiulianCard — per ui-spec/Card.spec.md.
///
/// Surface container with hairline border, three padding sizes, optional header / footer slots,
/// and optional interactive variant (tappable with hover lift + shadow).
public struct LiulianCard<Header: View, Content: View, Footer: View>: View {
    public enum Variant {
        case `static`
        case interactive
    }

    public enum Size {
        case compact
        case `default`
        case spacious
    }

    let variant: Variant
    let size: Size
    let selected: Bool
    let disabled: Bool
    let onPress: (() -> Void)?
    let header: () -> Header
    let content: () -> Content
    let footer: () -> Footer

    @State private var isHovered: Bool = false
    @State private var isPressed: Bool = false
    @FocusState private var isFocused: Bool

    public init(
        variant: Variant = .static,
        size: Size = .default,
        selected: Bool = false,
        disabled: Bool = false,
        onPress: (() -> Void)? = nil,
        @ViewBuilder header: @escaping () -> Header = { EmptyView() },
        @ViewBuilder content: @escaping () -> Content,
        @ViewBuilder footer: @escaping () -> Footer = { EmptyView() }
    ) {
        self.variant = variant
        self.size = size
        self.selected = selected
        self.disabled = disabled
        self.onPress = onPress
        self.header = header
        self.content = content
        self.footer = footer
    }

    public var body: some View {
        let inner = VStack(alignment: .leading, spacing: gap) {
            if !(header() is EmptyView) {
                header()
                Divider().background(LiulianTokens.Colors.hairline)
            }
            content()
            if !(footer() is EmptyView) {
                Divider().background(LiulianTokens.Colors.hairline)
                footer()
            }
        }
        .padding(padding)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(LiulianTokens.Colors.surfacePure)
        .clipShape(RoundedRectangle(cornerRadius: LiulianTokens.Radius.md))
        .overlay(
            RoundedRectangle(cornerRadius: LiulianTokens.Radius.md)
                .strokeBorder(borderColor, lineWidth: borderWidth)
        )
        .offset(y: isHovered && !isPressed ? -2 : 0)
        .scaleEffect(isPressed ? 0.99 : 1.0)
        .shadow(
            color: isHovered ? Color.black.opacity(0.04) : Color.clear,
            radius: 8, x: 0, y: 2
        )
        .opacity(disabled ? LiulianTokens.Opacity.disabled : 1.0)
        .overlay(
            RoundedRectangle(cornerRadius: LiulianTokens.Radius.md)
                .strokeBorder(
                    isFocused ? LiulianTokens.Colors.unibeRed : Color.clear,
                    lineWidth: LiulianTokens.Focus.ringWidth
                )
                .padding(-(LiulianTokens.Focus.ringOffset))
        )
        .animation(
            .timingCurve(0.16, 1.0, 0.3, 1.0, duration: LiulianTokens.Duration.medium),
            value: isHovered
        )

        if variant == .interactive, let onPress = onPress {
            Button(action: { if !disabled { onPress() } }) { inner }
                .buttonStyle(.plain)
                .focused($isFocused)
                .disabled(disabled)
                .onLongPressGesture(
                    minimumDuration: .infinity,
                    maximumDistance: .infinity,
                    perform: {},
                    onPressingChanged: { pressing in isPressed = pressing }
                )
#if os(macOS)
                .onHover { hovering in isHovered = hovering }
#endif
        } else {
            inner
        }
    }

    private var padding: CGFloat {
        switch size {
        case .compact:  return LiulianTokens.Spacing.s4
        case .default:  return LiulianTokens.Spacing.s6
        case .spacious: return LiulianTokens.Spacing.s7
        }
    }

    private var gap: CGFloat {
        switch size {
        case .compact:  return LiulianTokens.Spacing.s3
        case .default:  return LiulianTokens.Spacing.s4
        case .spacious: return LiulianTokens.Spacing.s5
        }
    }

    private var borderColor: Color {
        if selected { return LiulianTokens.Colors.unibeRed }
        if isHovered { return LiulianTokens.Colors.hairlineStrong }
        return LiulianTokens.Colors.hairline
    }

    private var borderWidth: CGFloat {
        selected ? 2 : 1
    }
}

// MARK: - Preview

#if DEBUG
struct LiulianCard_Previews: PreviewProvider {
    static var previews: some View {
        VStack(spacing: LiulianTokens.Spacing.s4) {
            LiulianCard {
                LiulianText("Aare-Bern", variant: .title)
                LiulianText("Water level 142 cm · 24h forecast +6 cm", variant: .body, color: LiulianTokens.Colors.inkMuted)
            }

            LiulianCard(variant: .interactive, onPress: {}) {
                LiulianText("Forecast canvas", variant: .title)
                LiulianText("Tap to view the full forecast workspace.", variant: .body, color: LiulianTokens.Colors.inkMuted)
            }

            LiulianCard(selected: true) {
                LiulianText("Selected station", variant: .title)
                LiulianText("Border thickens, color shifts to UniBe red.", variant: .body, color: LiulianTokens.Colors.inkMuted)
            }
        }
        .padding(LiulianTokens.Spacing.s6)
        .background(LiulianTokens.Colors.canvasWarm)
    }
}
#endif

File: android/liulian-ui/src/main/kotlin/io/liulian/ui/LiulianCard.kt
package io.liulian.ui

import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import io.liulian.tokens.LiulianTokens

enum class CardSize { Compact, Default, Spacious }

/**
 * LiulianCard — per ui-spec/Card.spec.md.
 *
 * Surface container with hairline border, three padding sizes, optional onPress (interactive).
 */
@Composable
fun LiulianCard(
    modifier: Modifier = Modifier,
    size: CardSize = CardSize.Default,
    interactive: Boolean = false,
    selected: Boolean = false,
    disabled: Boolean = false,
    onPress: (() -> Unit)? = null,
    header: (@Composable () -> Unit)? = null,
    footer: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()

    val lift by animateDpAsState(
        targetValue = if (interactive && !pressed && !disabled) 0.dp else 0.dp,
        animationSpec =
            tween(
                durationMillis = LiulianTokens.Duration.medium,
                easing = LiulianTokens.Easing.easeOutQuart,
            ),
        label = "cardLift",
    )

    val padding =
        when (size) {
            CardSize.Compact -> LiulianTokens.Spacing.s4
            CardSize.Default -> LiulianTokens.Spacing.s6
            CardSize.Spacious -> LiulianTokens.Spacing.s7
        }
    val gap =
        when (size) {
            CardSize.Compact -> LiulianTokens.Spacing.s3
            CardSize.Default -> LiulianTokens.Spacing.s4
            CardSize.Spacious -> LiulianTokens.Spacing.s5
        }

    val borderColor: Color = if (selected) LiulianTokens.Colors.unibeRed else LiulianTokens.Colors.hairline
    val borderWidth = if (selected) 2.dp else 1.dp

    var rootModifier =
        modifier
            .fillMaxWidth()
            .alpha(if (disabled) LiulianTokens.Opacity.disabled else 1f)
            .offset(y = lift)
            .shadow(elevation = if (pressed) 0.dp else 0.dp, shape = RoundedCornerShape(LiulianTokens.Radius.md))
            .clip(RoundedCornerShape(LiulianTokens.Radius.md))
            .background(LiulianTokens.Colors.surfacePure)
            .border(borderWidth, borderColor, RoundedCornerShape(LiulianTokens.Radius.md))

    if (interactive && onPress != null) {
        rootModifier =
            rootModifier.clickable(
                interactionSource = interactionSource,
                indication = null,
                enabled = !disabled,
                onClick = onPress,
            )
    }
    rootModifier = rootModifier.padding(padding)

    Column(
        modifier = rootModifier,
        verticalArrangement = Arrangement.spacedBy(gap),
    ) {
        if (header != null) {
            header()
            Box(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(LiulianTokens.Colors.hairline),
            )
        }
        content()
        if (footer != null) {
            Box(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(LiulianTokens.Colors.hairline),
            )
            footer()
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFBFBFA)
@Composable
private fun LiulianCardPreview() {
    Column(
        modifier =
            Modifier
                .padding(LiulianTokens.Spacing.s6)
                .background(LiulianTokens.Colors.canvasWarm),
        verticalArrangement = Arrangement.spacedBy(LiulianTokens.Spacing.s4),
    ) {
        LiulianCard(size = CardSize.Compact) {
            LiulianText("Compact", variant = LiulianTextVariant.Title)
            LiulianText("padding s4 (16)", variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkMuted)
        }
        LiulianCard {
            LiulianText("Default", variant = LiulianTextVariant.Title)
            LiulianText("padding s6 (24)", variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkMuted)
        }
        LiulianCard(size = CardSize.Spacious) {
            LiulianText("Spacious", variant = LiulianTextVariant.Title)
            LiulianText("padding s7 (32)", variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkMuted)
        }
        LiulianCard(interactive = true, onPress = {}) {
            LiulianText("Interactive", variant = LiulianTextVariant.Title)
            LiulianText("press = scale 0.99", variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkMuted)
        }
        LiulianCard(selected = true) {
            LiulianText("Selected", variant = LiulianTextVariant.Title)
            LiulianText("border 2dp unibeRed", variant = LiulianTextVariant.Body, color = LiulianTokens.Colors.inkMuted)
        }
    }
}
